src/feature/hs/hs_common.c

```diff
@@ -1723,45 +1716,41 @@ hs_get_extend_info_from_lspecs(const smartlist_t *lspecs,
     }
   } SMARTLIST_FOREACH_END(ls);
 
-  /* Legacy ID is mandatory, and we require IPv4. */
-  if (!have_v4 || !have_legacy_id) {
+  /* Choose a preferred address first, but fall back to an allowed address.
+   * Skip IPv6 if we're not direct as relays only extend via IPv4. */
+  fascist_firewall_choose_address_ls(lspecs, 0, &ap, direct_conn);
+
+  /* Legacy ID is mandatory, and we require an IP address. */
+  if (!tor_addr_port_is_valid_ap(&ap, 0) || !have_legacy_id) {
+    /* If we're missing the legacy ID or the IP address, return NULL. */
     goto done;
   }
 
-  /* We know we have IPv4, because we just checked. */
+  /* We know we have an IP address, because we just checked. */
   if (!direct_conn) {
```

Comment:
Please remove this entire if...else block, and the "validate:" label. It is redundant.

But keep the comment about falling back to a 3-hop path.